src/utils/ucc_compiler_def.h

```diff
@@ -25,6 +25,11 @@
 #define ucc_likely        ucs_likely
 #define ucc_unlikely      ucs_unlikely
 
+/**
+ * Prevent compiler from reordering instructions
+ */
+#define ucc_compiler_fence()       asm volatile(""::: "memory")
```

Comment:
@Sergei-Lebedev i think i was not right saying that Shimmy needs just a compiler instruction barrier. We need arch specific CPU instruction barrier.
in hcoll we use libocoms which defines isync:
power: https://github.com/uccs/libocoms/blob/4aa64d0105317badb244e7433ea29aa03a014507/ocoms/sys/powerpc/atomic.h#L34

arm: https://github.com/uccs/libocoms/blob/4aa64d0105317badb244e7433ea29aa03a014507/ocoms/sys/arm64/atomic.h#L76

@bureddy we need an instruction sync primitive for shared memory collectives (remember ppc data race) - does ucx/ucs have that or we need to copy paste from libocoms?

Reply:
ucx/ucs  ucc_memory_cpu_load_fence() should be enough for pcc shared memory.

Reply:
here was the requirement  on PPC
```

producer 
//store data
for( .............)
 shared_array[i] = 0;                                                                                                                  
ppc_write_fence();   //an lwsync is sufficient for DDR memory writes
//Notify consumers
shared_flag = 1;

Consumer
while (!shared_flag);    //blocking wait on the shared flag (ideally should be non-blocking)
isync();                             //Don't let the speculation unit move instructions ahead of the while loop check
//access the shared_array
int x = shared_array[y];     //read the updated shared state

```